monix-eval/shared/src/test/scala/monix/eval/TaskMemoizeSuite.scala

```diff
@@ -663,4 +665,39 @@ object TaskMemoizeSuite extends BaseTestSuite {
       assertEquals(v, (100, 100))
     }
   }
+
+  testAsync("Task.memoize doesn't corrupt Local (issue #856)") { _ =>
+    import monix.execution.Scheduler.Implicits.global
+    implicit val opts = Task.defaultOptions.enableLocalContextPropagation
+
+    val memoizedTask = Task(5).memoize
+
+    val i = AtomicInt(0)
+
+    val t = for {
+      local <- TaskLocal(0)
+      ii    <- Task.evalAsync(i.incrementAndGet())
+      _     <- local.write(ii)
+      result <- Task.parZip2(
+        memoizedTask.flatMap { _ =>
+          local.read.executeAsync
+        },
+        memoizedTask.flatMap { _ =>
+          local.read.executeAsync
+        }
+      )
+    } yield result
+
+    val f1 = t.runToFutureOpt
+    val f2 = t.runToFutureOpt
+    val f3 = t.runToFutureOpt
```

Comment:
BTW, one of the crazy Scala features is that you can write:
`val f1, f2, f3 = t.runToFutureOpt`

Not sure if you want to, though 😄